tools/chapel-py/method-tables.h

```diff
@@ -0,0 +1,461 @@
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// This file defines nearly all the methods exposed to python from the frontend.
+//
+// To use these definitions, define the macro and then `#include` this file.
+// See `chapel.cpp` for examples. If the macros below are not defined, they are
+// defined to be empty. Only the ones that are used at the time of include need
+// to be defined.
+//
+
+//
+// Defines the beginning of an ast node
+//
+#ifndef METHOD_TABLE_BEGIN
+#define METHOD_TABLE_BEGIN(TAG)
+#endif
+
+//
+// Defines the end of an ast node
+//
+#ifndef METHOD_TABLE_END
+#define METHOD_TABLE_END(TAG)
+#endif
+
+//
+// Defines a simple getter that takes no arguments
+//
+// BODY is an inline definition of the getter. Available variables:
+// - contextObject: the current frontend context
+// - node: the ast node
+// TYPESTR is the return type. Possible values:
+// - "O" = python object
+// - "s" = string
+// - "b" = bool
+//
+#ifndef PLAIN_GETTER
+#define PLAIN_GETTER(NODE, NAME, DOCSTR, TYPESTR, BODY)
+#endif
+
+//
+// Declares a simple getter that takes no arguments. The definition is defined
+// elsewhere in a function named `NODEObject_NAME`. See the `ACTUAL_ITERATOR`
+// macro for an example
+//
+#ifndef METHOD_PROTOTYPE
+#define METHOD_PROTOTYPE(NODE, NAME, DOCSTR)
+#endif
+
+
+//
+// The order here should be kept in sync with the order in the uast list.
+// Not all nodes need methods exposed to python, so not all uast nodes are listed here.
+// See list in frontend/include/chpl/uast/uast-classes-list.h.
+//
+// Inside each method table, methods should be listed in alphabetical order
+//
+
+METHOD_TABLE_BEGIN(AnonFormal)
+  PLAIN_GETTER(AnonFormal, intent, "Get the intent for this formal",
+               "s", return intentToString(node->intent()))
+  PLAIN_GETTER(AnonFormal, type_expression, "Get the type expression for this formal",
+               "O", return  wrapAstNode(contextObject, node->typeExpression()))
+METHOD_TABLE_END(AnonFormal)
+
+METHOD_TABLE_BEGIN(Array)
+  PLAIN_GETTER(Array, exprs, "Get the Array exprs",
+               "O", return wrapIterPair(contextObject, node->exprs()))
+  PLAIN_GETTER(Array, has_trailing_comma, "Check if the Array had a trailing comma",
+               "b", return node->hasTrailingComma())
+  PLAIN_GETTER(Array, is_associative, "Check if the Array is associative",
+               "b", return node->isAssociative())
```

Comment:
I'm only now reading the doc strings for some of these (because the layout is so much nicer!) and I wanted to point out that my convention has been to capitalize node names (so write "Array" when saying "Array node"), but normally case language constructs ("string", "array"). Also, I try not to use abbreviations I think -- "expressions" instead of "exprs". Perhaps: "Get the expressions from this Array node"? Similar updates could be done (but I won't hold you to it) to other methods/classes onwards.